File: data/spark/code/python/ml/recommendation/ALS.py
from pyspark.ml.recommendation import ALS
from pyspark.sql import SparkSession
from pyspark.sql import Rows
from pyspark.sql import functions

spark = SparkSession.builder.appName("MyApp").getOrCreate()
rdd = spark.read.text("file.txt").rdd
df = spark.createDataFrame(rdd)

# ALS (Alternating Least Squares)
# this is a Machine Learning recommendation algorithm
estimator = ALS(maxIter=5,
                regParam=0.01,
                userCol="userID",
                itemCol="movieID",
                ratingCol="rating")

# trains a model with a dataframe
model = estimator.fit(df)

# run recommendation model for a set of movies that you'd like to recommend
recommendations = model.transform(recommendedMoviesDf)

# get top 20 recommendations
topRecommendations = recommendations.sort(
    recommendations.prediction.desc()).take(20)

File: data/Storage/Spark/code/scala/src/main/scala/com/hv/spark/ml/recommendation/ALS.scala
package com.hv.spark.ALS

import org.apache.log4j.Level
import org.apache.log4j.Logger
import org.apache.spark.SparkContext
import org.apache.spark.broadcast.Broadcast
import org.apache.spark.ml.recommendation.ALS
import org.apache.spark.ml.recommendation.ALSModel
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.SparkSession
import org.apache.spark.util.LongAccumulator

// Alternating Least Squares

object Main {
  def main(args: Array[String]) {
    Logger.getLogger("org").setLevel(Level.ERROR)
    val df = InitDF.run()

    // Static Methods
    val als = ALSNew.run()

    // Instance Methods
    ALSFit.run(als, df)

  }
}

object InitDF {
  def run() = {
    val ss: SparkSession = SparkSession
      .builder()
      .appName("MyApp")
      .master("local[*]")
      .getOrCreate()
    import ss.implicits._
    val df: DataFrame = ss.read
      .option("header", "true")
      .option("inferSchema", "true")
      .csv("ml-latest-small/movies.csv")
    df
  }
}

object ALSNew {
  def run() = {
    val als: ALS = new ALS()
      .setMaxIter(5)
      .setRegParam(0.01)
      .setUserCol("userID")
      .setItemCol("movieId")
      .setRatingCol("rating")
    als
  }
}

object ALSFit {
  def run(als: ALS, df: DataFrame) = {
    val model: ALSModel = als.fit(df)
  }
}
